Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -1212,76 +1386,100 @@ std::tuple<double, double> DiscusMultipleScatteringCorrection::getKinematicRange
- * @param scatteringXSection The scatter cross section applicable to this scatter
- * @param rng Random number generator
- * @param QSS Sum of Q.S(Q) - used to normalise when not importance sampling
- * @param weight The weight for this track
- */
-void DiscusMultipleScatteringCorrection::q_dir(Geometry::Track &track, const MatrixWorkspace_sptr &invPOfQ, double &k,
+// update track direction and weight
+bool DiscusMultipleScatteringCorrection::q_dir(Geometry::Track &track, const Geometry::IObject *shapePtr,
+                                               const ComponentWorkspaceMappings &componentWorkspaces, double &k,
                                                const double scatteringXSection,
-                                               Kernel::PseudoRandomNumberGenerator &rng, double &QSS, double &weight) {
+                                               Kernel::PseudoRandomNumberGenerator &rng, double &weight) {
   const double kinc = k;
-  double QQ, SQ;
+  double QQ;
   int iW;
+  auto componentWSIt = findMatchingComponent(componentWorkspaces, shapePtr);
   if (m_importanceSampling) {
-    std::tie(QQ, iW) = sampleQW(invPOfQ, rng.nextValue());
-    k = getKf(m_SQWS->getAxis(1)->getValue(iW), kinc);
-    // S(Q) not strictly needed here but useful to see if the higher values are indeed being returned
-    SQ = interpolateFlat(m_SQWS->getSpectrum(iW), QQ);
+    std::tie(QQ, iW) = sampleQW(componentWSIt->InvPOfQ, rng.nextValue());
+    k = getKf(componentWSIt->SQ->getAxis(1)->getValue(iW), kinc);
     weight = weight * scatteringXSection;
   } else {
-    auto wAxis = dynamic_cast<NumericAxis *>(m_logSQ->getAxis(1));
-    if (!wAxis)
-      throw std::invalid_argument("Cannot sample w on S(Q,w) without a numeric w axis");
-    std::tie(k, iW) = sampleKW(wAxis->getValues(), rng, kinc);
-    auto [qmin, qrange] = getKinematicRange(k, kinc);
-    QQ = qmin + qrange * rng.nextValue();
-    SQ = interpolateGaussian(m_logSQ->getSpectrum(iW), QQ);
-    weight = weight * scatteringXSection * SQ * QQ;
+    auto &wValues = dynamic_cast<NumericAxis *>(componentWSIt->SQ->getAxis(1))->getValues();
+    // in order to keep integration limits constant sample full range of w even if some not kinematically accessible
+    // Note - Discus took different approach where it sampled q,w from kinematically accessible range only but it
+    // only calculated for double scattering and easier to normalise in that case
+    double wRange;
+    /*// the energy transfer must always be less than the positive value corresponding to energy going from ki to 0
+    // Note - this is still the case for indirect because on a multiple scatter the kf isn't kfixed
+    double wMax = fromWaveVector(kinc);
+    // find largest w bin centre that is < wmax and then sample w up to the next bin edge
+    auto it = std::lower_bound(wValues.begin(), wValues.end(), wMax);
+    int iWMax = static_cast<int>(std::distance(wValues.begin(), it) - 1);*/
```

Comment:
Yes they are but the code would benefit from a further comment here about why. I'll add one in. The code that's commented out here and also on L1452 is an optimisation that should mean you can get more accurate results but I didn't want to introduce it at this stage because I've spent a long time on validating the results without this. So in the spirit of taking things one step at a time I was planning on raising a separate PR for this